File: Python/permutation_no_duplication(int).py
def permute(arr, left, right):
    if left == right:
        print(arr)
    else:
        for i in range(left, right + 1):
            should_swap = True

            # Check for duplicates and avoid swapping if duplicates are found
            for j in range(left, i):
                if arr[j] == arr[i]:
                    should_swap = False
                    break

            if should_swap:
                arr[left], arr[i] = arr[i], arr[left]
                permute(arr, left + 1, right)
                arr[left], arr[i] = arr[i], arr[left]  # Backtrack

def generate_permutations(arr):
    n = len(arr)

    print(f"Permutations of {arr} with duplicates:")
    permute(arr, 0, n - 1)

if __name__ == "__main__":
    input_array = [1, 2, 2]
    generate_permutations(input_array)

File: C/combinatoric/permutation_no_duplication(int).c
#include <stdio.h>

void swap(int *x, int *y) {
    int temp;
    temp = *x;
    *x = *y;
    *y = temp;
}

int containsDuplicate(int arr[], int left, int right, int key) {
    for (int i = left; i <= right; i++) {
        if (arr[i] == key) {
            return 1; // Duplicate found
        }
    }
    return 0; // No duplicate
}

void permute(int arr[], int left, int right) {
    if (left == right) {
        for (int i = 0; i <= right; i++) {
            printf("%d ", arr[i]);
        }
        printf("\n");
    } else {
        for (int i = left; i <= right; i++) {
            if (!containsDuplicate(arr, left, i - 1, arr[i])) {
                swap(&arr[left], &arr[i]);
                permute(arr, left + 1, right);
                swap(&arr[left], &arr[i]); // Backtrack
            }
        }
    }
}

int main() {
    int arr[] = {1, 2, 2};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Permutations of the array with duplicates:\n");
    permute(arr, 0, n - 1);

    return 0;
}
